cunumeric/deferred.py

```diff
@@ -1132,6 +1132,40 @@ def dot(self, src1, src2, stacklevel=0, callsite=None):
                 f"dot between {rhs1_array.ndim}d and {rhs2_array.ndim}d arrays"
             )
 
+    # Create array from input array and indices
+    @profile
+    @auto_convert([1])
+    @shadow_debug("choose", [1])
+    def choose(
+        self,
+        *args,
+        choices,
+        out,
+        is_tuple,
+        mode="rase",
+        stacklevel=0,
+        callsite=None,
+    ):
+        out_arr = out.base
+        index_arr = self._broadcast(out_arr.shape)
+        ch_tuple = tuple()
+        for c in args:
+            c = c._broadcast(out_arr.shape)
+            ch_tuple = ch_tuple + (c,)
+        task = self.context.create_task(CuNumericOpCode.CHOOSE)
+        task.add_output(out_arr)
+        task.add_input(index_arr)
+        if not is_tuple:
+            task.add_input(choices.base)
+            task.add_broadcast(choices.base)
```

Comment:
Do you really need to broadcast all choices arrays in full to every point task? I would assume that, e.g., if `a.shape == (10,)`, `forall i: choices[i].shape == (10,)` and you have two processors, then processor A would get the first 5 element from all arrays, and processor B would get the other 5 elements. No processor would need to get the entire 10 elements from any array (which is what `add_broadcast` does).

Reply:
Yes, I believe I do since access to choices is very sparse and defined by the input array. So it is not possible to predict which "subarray" in choices will be accessed in each index point.

Reply:
I guess this doesn't quite work if the choices are all passed inside the same array, but this won't be an issue if you convert them to a tuple at the beginning of the array.py choose function.

Reply:
Could you, please, clarify "this doesn't quite work if the choices are all passed inside the same array"? 
I would still need to broadcast all arrays in the tuple.

Reply:
> Yes, I believe I do since access to choices is very sparse and defined by the input array. So it is not possible to predict which "subarray" in choices will be accessed in each index point.

I don't think this is the case. Imagine you're trying to distribute this call across 2 processors:

```
np.choose( a = (0,1,0,1,0,1), choices = ([1,2,3,4,5,6], [-1,-2,-3,-4,-5,-6]) )
```

I posit that processor A only needs the first three elements of each of `a`, `choices[0]` and `choices[1]`, and processor B only needs the last three elements from each array. Then each processor can proceed independently, and fill out a separate subset of the output array. No processor needs access to any of the arrays in full.

Is this analysis inaccurate?

Reply:
I see, so you suggest passing partitioned subarrays. This makes sense.